platform/security/core/security-core-impl/src/main/java/ddf/security/common/util/SecurityUtils.java

```diff
@@ -0,0 +1,217 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SecurityUtils.class);
+
+    private static Subject cachedSystemSubject;
+
+    private static Object lock = new Object();
+
+    public SecurityUtils() {
+
+    }
+
+    public Subject getSubject(String username, String password) {
+        UPAuthenticationToken token = new UPAuthenticationToken(username, password);
+        SecurityManager securityManager = getSecurityManager();
+        if (securityManager != null) {
+            try {
+                return securityManager.getSubject(token);
+            } catch (Exception sse) {
+                LOGGER.error("Unable to request subject for {} user.", username, sse);
+            }
+        }
+        return null;
+    }
+
+    public boolean tokenAboutToExpire(Subject subject) {
+        boolean tokenAboutToExpire = true;
+        if ((null != subject) && (null != subject.getPrincipals()) && (null != subject
+                .getPrincipals().oneByType(SecurityAssertion.class)) && (!subject.getPrincipals()
+                .oneByType(SecurityAssertion.class).getSecurityToken()
+                .isAboutToExpire(TimeUnit.MINUTES.toSeconds(1)))) {
+            tokenAboutToExpire = false;
+        }
+        return tokenAboutToExpire;
+    }
+
+    public boolean javaSubjectHasAdminRole() {
+        javax.security.auth.Subject subject = javax.security.auth.Subject
+                .getSubject(AccessController.getContext());
+        if (subject != null) {
+            return subject.getPrincipals().contains(ADMIN_ROLE);
+        }
+        return false;
+    }
+
+    public Subject getSystemSubject() {
+
+        if (!tokenAboutToExpire(cachedSystemSubject)) {
+            return cachedSystemSubject;
+        }
+
+        KeyStore keyStore = getSystemKeyStore();
+        String alias = null;
+        Certificate cert;
+        try {
+            if (keyStore.size() == 1) {
+                alias = keyStore.aliases().nextElement();
+            } else if (keyStore.size() > 1) {
+                alias = getCertificatAlias();
+            }
+            cert = keyStore.getCertificate(alias);
+        } catch (KeyStoreException e) {
+            LOGGER.error("Unable to get certificate for alias [{}]", alias, e);
+            return null;
+        }
+
+        if (cert == null) {
+            LOGGER.error("Unable to get certificate for alias [{}]", alias);
+            return null;
+        }
+
+        PKIAuthenticationTokenFactory pkiTokenFactory = createPKITokenFactory();
+        PKIAuthenticationToken pkiToken = pkiTokenFactory
+                .getTokenFromCerts(new X509Certificate[] {(X509Certificate) cert},
+                        PKIAuthenticationToken.DEFAULT_REALM);
+        if (pkiToken != null) {
+            SecurityManager securityManager = getSecurityManager();
+            if (securityManager != null) {
+                synchronized (lock) {
```

Comment:
You can just make this method synchronized for simplicity instead of locking this critical section.